LFS-199: As an administrator, I can index vocabularies from OBO sources

prabnur opened the pull request.

prabnur committed:
LFS-199: As an administrator, I can index vocabularies from OBO sources
Initial version

prabnur commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OboParser.java:
```diff
@@ -0,0 +1,251 @@
+        this.crtTerm.addTo(name, value.replaceFirst("^\"(.+)\"\\s*?(?:[A-Z]+|\\[).*", "$1")
+            .replaceFirst("\\s+\\{.*$", "")
+            .replaceFirst("^(HP:\\d{7}) ! .*$", "$1")
+            .replace("\\\"", "\""));
```
For this one, I decided to use examples to show how the string gets changed.

prabnur commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OboParser.java:
```diff
@@ -0,0 +1,251 @@
+                return ancestors;
+            } else {
+                Collection<String> singleElement = new LinkedHashSet<>();
+                singleElement.add(termID);
```
Yes because the return mechanism is for the ancestor to tell its children it is one. It doesn't set anything or make additional calls for the root node.

prabnur replied:
I changed the name to rootElement so its role is more clear

veronikaslc commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/TermData.java:
```diff
@@ -0,0 +1,161 @@
+    /**
+     * Returns an empty Collection.
+     */
+    protected Collection<String> getEmptyCollection()
```
I'm not entirely sure we need a method that just returns empty collection. It is used only once and do not hold much of a code in self. I'd rather just `return new LinkedHashSet<>();` whenever needed in other methods.

marta- replied:
@prabnur https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#emptySet--

veronikaslc replied:
Right, just `return Collections.emptySet();`

prabnur commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OboParser.java:
```diff
@@ -0,0 +1,251 @@
+                Collection<String> parents = term.getCollection(TermData.PARENT_FIELD_NAME);
+                for (String parent : parents) {
+                    //this.logger.info("Current " + termID + "; Parent " + parent);
+                    ancestors.addAll(findAncestors(parent));
```
Can I talk to you about this one?

prabnur commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OboParser.java:
```diff
@@ -0,0 +1,251 @@
+        if (term.hasKey(TermData.TERM_CATEGORY_FIELD_NAME)) {
+            return term.getCollection(TermData.TERM_CATEGORY_FIELD_NAME);
+        } else {
+            if (term.hasKey(TermData.PARENT_FIELD_NAME)) {
```
I added comments to explain it better. All parents are ancestors. I can go into more detials, but I think my comments do  a decent job of that now.

veronikaslc commented on modules/vocabularies/src/test/java/ca/sickkids/ccm/lfs/vocabularies/VocabularyIndexerServletTest.java:
```diff
@@ -476,4 +485,111 @@ private void ncitFlatTestC100005(Node flatTestVocabulary)
+        try {
+            Assert.assertTrue(oboParser.canParse("OBO"));
+            oboParser.parse(temporaryFile, description, new OboConsumerTest());
+        } catch (Exception e) {
```
Test runs successfully no matter what even with non-existent test file name. The reason is that all exceptions are already caught in oboParser.parse() and nothing gets thrown.

prabnur replied:
Ohhh noooooo

veronikaslc replied:
At least that what I get running test file as JUnit test in debug mode in Eclipse

veronikaslc commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OboParser.java:
```diff
@@ -0,0 +1,233 @@
+@Component(
+    service = SourceParser.class,
+    name = "SourceParser.OBO")
+@SuppressWarnings("checkstyle:ClassFanOutComplexity")
```
This line should be removed, nothing to suppress. Default maximum threshold for ClassFanOutComplexity is 20.

veronikaslc commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OboParser.java:
```diff
@@ -0,0 +1,233 @@
+    private void consumeData(final Consumer<VocabularyTermSource> consumer)
+    {
+        String[] typeString = {};
+        String sample = "";
```
I'm not sure I get the need of this `sample` variable. Is it used?

marta- commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OboParser.java:
```diff
@@ -0,0 +1,249 @@
+     * @return IDs of the ancestors of the given term
+     * @throws NullPointerException if there was no term in the source file with the ID termID
+     */
+    private Collection<String> findAncestors(String termID)
```
This method has inconsistent behavior: for all non-roots it returns the ancestors, while for the root it returns the node itself. Moreover, the ancestors are never set for roots (should just be the empty list) so every time the root is reached the ancestors are re-computed.

Here's a better logic for this function

```
if the ancestors for this node have already been determined {
  return them as a list of IDs
} else {
  if the term has parents {
    parents = the term's parents;
  } else {
    parents = empty set;
  }
  ancestors = set initialized with parents;
  for each parent {
     ancestors.addAll(findAncestors(parent))
  }
  return ancestors;
}
```
Note how the roots are treated just like every other node, and recursive calls naturally stop at the root because the parents set is empty so the for loop doesn't run.

prabnur replied:
That makes sense!

marta- commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/TermData.java:
```diff
@@ -0,0 +1,151 @@
+    public static final String PARENT_FIELD_NAME = "is_a";
+
+    /** Regex Pattern matching a Parent's ID and description. */
+    public static final String PARENT_ID_REGEX = "^([A-Z]+\\:[0-9]{7})\\s*!\\s*.*";
```
Not all IDs have 7 digits. For example, CHEBI has 5, SNOMED has 9, UMLS and MSH have a lcapital letter after `:`.

prabnur replied:
Is there a more complete description of these? I want to make sure I don't want to miss anything.

marta- replied:
None that I know of. I would suggest relying on the obo stanza structure when extracting the parent, rather than trying to recognize the id: https://owlcollab.github.io/oboformat/doc/GO.format.obo-1_2.html#S.1.3

marta- commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/OboParser.java:
```diff
@@ -0,0 +1,249 @@
+                // "  {xref="PMID:16424154"}" at the end of the String -> ""
+                .replaceFirst("\\s+\\{.*$", "")
+                // "HP:1234567 ! ..." at start of String -> "HP:1234567"
+                .replaceFirst("^(HP:\\d{7}) ! .*$", "$1")
```
This is still hardcoded for HPO.

sdumitriu committed:
LFS-199: As an administrator, I can index vocabularies from OBO sources
Bulletproofing the parser
- add support for line continuations
- add support for joining multiple frames for the same term
- add support for escape sequences
- multiple duplicate values are coalesced
- fixed support for quoted strings
- fixed potential resource leak due to unclosed file
- fixed bug when last frame isn't a term frame
- fixed possible NPE when a specified parent doesn't actually exist
- add missing implementation for VocabularyTermSource.getDescription
- fixed and improved comments
- better and more comprehensive tests

sdumitriu committed:
[cleanup] Fixed wrong javadoc

sdumitriu committed:
[misc] Re-enable tests

sdumitriu committed:
[misc] Moved test resource in the right place

marta- committed:
Apply suggestions from code review

Co-Authored-By: OrdiNeu <OrdinaryNeumann@gmail.com>

OrdiNeu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/TermData.java:
```diff
@@ -0,0 +1,148 @@
+     *
+     * @param key - for value to be assigned.
+     * @param value - that is to be assigned to key.
+     * @return whether or not the add was successful.
```
Note for posterity: this isn't technically whether the add was successful. [This returns whether the add has changed the map](https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/multimap/AbstractMultiValuedMap.html#put-K-V-) (and will return false if you add the same value twice).

sdumitriu committed:
LFS-199: As an administrator, I can index vocabularies from OBO sources
Better javadoc
Better test function names
